File: backend/schemas/credentials.py
"""Pydantic schemas for the Credentials domain.

Mirrors :mod:`backend.db.models.credentials.Credential`. Schemas are
intentionally minimal — credentials are admin-only, ``ri``-gated data
that never ships to public APIs.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class CredentialCreate(BaseModel):
    """Payload to create a new credentials entry.

    The caller supplies a title and the desired filename (relative to
    ``settings.credentials_storage_path``); the service writes the
    initial content to disk and inserts the registry row.
    """

    title: str = Field(..., min_length=1, max_length=500)
    filename: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description=("Filename relative to credentials_storage_path. Must not contain slashes — flat directory only."),
    )
    content: str = Field(default="", description="Initial file content (UTF-8 markdown).")


class CredentialUpdate(BaseModel):
    """Partial update — only the title is mutable; file_path is identity."""

    title: str | None = Field(default=None, min_length=1, max_length=500)


class CredentialRead(BaseModel):
    """Serialised credentials registry row."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    file_path: str
    created_at: datetime
    updated_at: datetime


class CredentialContent(BaseModel):
    """On-disk content of a credentials file."""

    model_config = ConfigDict(from_attributes=True)

    credential_id: UUID
    file_path: str
    content: str
    size_bytes: int = Field(..., ge=0)


class CredentialContentUpdate(BaseModel):
    """Payload for ``PUT /credentials/{id}/content`` — overwrite the file."""

    content: str = Field(..., description="New full file content (UTF-8 markdown).")

File: frontend/src/services/api/credentials.ts
import api from "../api";
import type {
  CredentialRead,
  CredentialCreate,
  CredentialUpdate,
  CredentialContent,
  CredentialContentUpdate,
} from "../../types/credential";

export function listCredentials(): Promise<CredentialRead[]> {
  return api.get<CredentialRead[]>("/credentials");
}

export function getCredential(id: string): Promise<CredentialRead> {
  return api.get<CredentialRead>(`/credentials/${id}`);
}

export function getCredentialContent(id: string): Promise<CredentialContent> {
  return api.get<CredentialContent>(`/credentials/${id}/content`);
}

export function putCredentialContent(
  id: string,
  payload: CredentialContentUpdate,
): Promise<CredentialContent> {
  return api.put<CredentialContent>(`/credentials/${id}/content`, payload);
}

export function createCredential(data: CredentialCreate): Promise<CredentialRead> {
  return api.post<CredentialRead>("/credentials", data);
}

export function updateCredential(
  id: string,
  data: CredentialUpdate,
): Promise<CredentialRead> {
  return api.patch<CredentialRead>(`/credentials/${id}`, data);
}

export function deleteCredential(id: string): Promise<void> {
  return api.delete(`/credentials/${id}`);
}
